internal/impl/aws/input_kinesis.go

```diff
@@ -593,53 +594,10 @@ func (k *kinesisReader) runBalancedShards() {
 				continue
 			}
 
-			// There were no unclaimed shards, let's look for a shard to steal.
-			selfClaims := len(clientClaims[k.clientID])
-			for clientID, claims := range clientClaims {
-				if clientID == k.clientID {
-					// Don't steal from ourself, we're not at that point yet.
-					continue
-				}
-
-				// This is an extremely naive "algorithm", we simply randomly
-				// iterate all other clients with shards and if any have two
-				// more shards than we do then it's fair game. Using two here
-				// so that we don't play hot potatoes with an odd shard.
-				if len(claims) > (selfClaims + 1) {
-					randomShard := claims[(rand.Int() % len(claims))].ShardID
-					k.log.Debugf(
-						"Attempting to steal stream '%v' shard '%v' from client '%v' as client '%v'\n",
-						streamID, randomShard, clientID, k.clientID,
-					)
-
-					sequence, err := k.checkpointer.Claim(k.ctx, streamID, randomShard, clientID)
-					if err != nil {
-						if k.ctx.Err() != nil {
-							return
-						}
-						if !errors.Is(err, ErrLeaseNotAcquired) {
-							k.log.Errorf("Failed to steal shard '%v': %v\n", randomShard, err)
-						}
-						k.log.Debugf(
-							"Aborting theft of stream '%v' shard '%v' from client '%v' as client '%v'\n",
-							streamID, randomShard, clientID, k.clientID,
-						)
-						continue
-					}
-
-					k.log.Debugf(
-						"Successfully stole stream '%v' shard '%v' from client '%v' as client '%v'\n",
-						streamID, randomShard, clientID, k.clientID,
-					)
-					wg.Add(1)
-					if err = k.runConsumer(&wg, streamID, randomShard, sequence); err != nil {
-						k.log.Errorf("Failed to start consumer: %v\n", err)
-					} else {
-						// If we successfully stole the shard then that's enough
-						// for now.
-						break
-					}
-				}
+			// Note: Stealing only happens if all shards have already been claimed
+			err = k.stealShard(&wg, streamID, clientClaims)
+			if !(err == nil || err == errShardNotStolen) {
```

Comment:
Could we please use `errors.Is(err, errShardNotStolen)` instead?

Also we still need to handle cases where error is not `errShardNotStolen`. For instance for `return k.ctx.Err()`

It's a bit confusing though as in both cases we should just return?!

Reply:
> It's a bit confusing though as in both cases we should just return?!
it does work though :) if there is no error OR  it is shard not stolen == true and then with the negation, it would be `false`
whereas if there is an error, it would be false and with the negation it would be true.

Anyways, i changed the conditional around, hopefully easier to read